csrc/scheduler/reduction.cpp

```diff
@@ -507,19 +507,224 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
+    }
+  }
+  rparams->multiple_reds_per_blk = hp.bdimx > 1 || hp.iter_unroll_factor > 1;
+
+  if (rparams->multiple_reds_per_blk) {
+    rparams->block_dim_iter_dom = ParallelType::TIDx;
+  }
+
+  rparams->grid_dim_iter_dom =
+      flip_grid ? ParallelType::BIDy : ParallelType::BIDx;
+  if (hp.gidim > (flip_grid ? scheduler_utils::y_grid_limit
+                            : scheduler_utils::x_grid_limit)) {
+    rparams->split_grid_dim_iter_dom_outer = true;
+    if (flip_grid) {
+      gdimy = scheduler_utils::y_grid_limit;
+    } else {
+      gdimx = scheduler_utils::x_grid_limit;
+    }
+  }
+
+  rparams->flip_grid = flip_grid;
+
+  if (rparams->cross_block_inner_reduction) {
+    if (rparams->block_dim_iter_dom == ParallelType::TIDx) {
+      rparams->block_dim_inner_reduction = ParallelType::TIDy;
+    } else {
+      rparams->block_dim_inner_reduction = ParallelType::TIDx;
+    }
+  }
+
+  rparams->unroll_factor_inner_reduction = hp.redu_unroll_factor;
+
+  rparams->unroll_factor_iter_dom = hp.iter_unroll_factor;
+  rparams->vectorize_iter_dom = hp.iter_unroll_factor > 1;
+
+  rparams->lparams = LaunchParams(
+      gdimx,
+      gdimy,
+      LaunchParams::UNINITIALIZED_VAL,
+      rparams->multiple_reds_per_blk ? hp.bdimx : hp.bdimy,
+      rparams->multiple_reds_per_blk ? hp.bdimy
+                                     : LaunchParams::UNINITIALIZED_VAL,
+      LaunchParams::UNINITIALIZED_VAL);
+
+  if (isDebugDumpEnabled(DebugDumpOption::SchedulerDebug)) {
+    debug() << hp.toString() << std::endl;
+    debug() << rparams->toString() << std::endl;
+  }
+  return rparams;
+}
+
+std::optional<outerReduHeuristicParas> maybeBlockOuterReduction(
+    int64_t vectorize_factor,
+    int64_t sm_count,
+    int64_t max_unroll,
+    int64_t max_threads_per_block,
+    int64_t n_tensor_outputs,
+    int64_t total_iteration_numel,
+    int64_t total_reduction_numel) {
+  outerReduHeuristicParas hp(total_iteration_numel, total_reduction_numel);
+
+  // Step-1, set iteration dim
+  // (1) start with bdimx = 8, gidim = 1, iter_unroll = 1
+  hp.bdimx = std::min(8L, hp.iDimAvail());
+  hp.gidim = 1;
+  hp.iter_unroll_factor = 1;
+
+  // (2) increase iter_unroll to its maximum following two rules:
+  // (2.1) ensure divisible split
+  // (2.2) leave enough blocks to saturate the device.
+  // Test on H100 shows the smallest iteration dim we can fully vectorize is
+  // 4800, which corresponds to 75 or 56.8 % of SMs. Here 50% is used.
+  const int64_t min_blocks_fully_vectorize = sm_count / 2;
+  int64_t max_iter_unroll = vectorize_factor;
+  // only allow this adjustment when there is only one output tensor
+  // or it's a small reduction where block reduction is enforced.
+  // When has multiple outputs, reduce vectorization leads to regression,
+  // see cpp benchmark of gelu backward which has 2 outputs ( does't seem like a
+  // common usage since thunder.jit generated fusion has only 1).
```

Comment:
regression is attached in the [doc](https://docs.google.com/document/d/12yBhqBHlZFKyxjgXqHjXxOfzZjsjXySHvwR0_8LsfZg/edit#heading=h.yailf9wrbndl).

Reply:
Sorry that I might not be reading this right.

> When has multiple outputs, reduce vectorization leads to regression

Is this specific to the *number of outputs*, or rather `size of output buffer`? Since I assume your thunder repro were only outputting the reduction output?